README.md
# envsettings

* Update django settings from environment variables by updating `sys.modules`
* Mostly using `ast.literal_eval` to convert string to python object.
* Tested by converting all default django settings to string and back to python object and then running django command.

## todo
* simplify code into one function because most of the things that this code is doing are probably not required
1. Usage:
    - in `myproject.settings`:
        ```python
        import os
        import sys
        import ast
        
        from djangobuk_envsettings import update_from_env
        
        update_from_env(
            sys.modules[__name__],
            # default prefix for all variables
            pre='DJANGO_',
            # settings that can be updated from env
            # by default all settings are allowed (this option overrides)
            allowed=[
                'SECRET_KEY',
                'SITE_ID',
            ],
            # optional
            # extra settings and their types (to be used with extra_allowed)
            extra_mapping={
                'DATABASE_PATH': ast.literal_eval,
            },
            # optional
            # extra settings that can be updated from env
            extra_allowed=[
                'DATABASE_PATH',
            ]
        )
        
        # nothing more required
        DATABASES = {
            'default': {
                'ENGINE': 'django.db.backends.sqlite3',
                'NAME': DATABASE_PATH,
            }
        }

       ```


CHANGELOG.MD
# 0.2.0
- django 4.1 support

tests/tests.py
import ast
import functools
import os
from unittest import TestCase

from django.conf import global_settings

from djangobuk_envsettings.utils import settings_to_dict, eval_settings, MAPPING
from django.core.management import call_command

BASE_DIR = os.path.dirname(os.path.realpath(__file__))


class ContextDecorator:

    def __call__(self, f):
        @functools.wraps(f)
        def decorated(*args, **kwargs):
            with self:
                return f(*args, **kwargs)
        return decorated


class temp_django_settings(ContextDecorator):

    def __enter__(self):
        import django.conf.global_settings
        self.old_settings = django.conf.global_settings
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        import django.conf.global_settings
        django.conf.global_settings = self.old_settings


class temp_environ(ContextDecorator):

    def __enter__(self):
        self.old_env = dict(os.environ.copy())
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        os.environ.clear()
        os.environ.update(self.old_env)


class TestMapping(TestCase):

    @classmethod
    def setUpClass(cls) -> None:
        cls.django_default_settings = {
            k: getattr(global_settings, k) for k in dir(global_settings) if
            not k.startswith('_') and
            k.upper() == k
        }

    @temp_environ()
    def test_all_settings_mapped(self):
        """
        Default Django settings have
        coresponding casting type.
        """
        for k in settings_to_dict(global_settings):
            self.assertIn(k, MAPPING)

    @temp_environ()
    def test_eval_extra_map(self):
        """
        Given extra mapping, settings typemap is correct.
        """
        settings = {
                'A': "1,2,3,4,5",
                'B': '["127.0.0.1", 55, 43]',
                'C': "a|b|c|d",
        }
        extra_map = {
                'A': lambda v: v.split(','),
                'C': lambda v: v.split('|'),
                'B': ast.literal_eval,
        }

        evaulted = eval_settings(
            settings=settings,
            mapping=extra_map,
        )

        for k, v in evaulted.items():
            if k == 'A':
                self.assertEqual(['1', '2', '3', '4', '5'], v)
            if k == 'B':
                self.assertEqual(['127.0.0.1', 55, 43], v)
            if k == 'C':
                self.assertEqual(["a", "b", "c", "d"], v)

    @temp_environ()
    def test_eval_with_default_settings(self):
        """
        Given default Django settings, convert them
        to string, cast them back and check their type.
        """
        settings_str = {
            k: str(v) for (k, v) in self.django_default_settings.items()
        }
        settings_eval = eval_settings(settings_str, mapping=MAPPING)

        self.assertCountEqual(settings_str, settings_eval)
        self.assertCountEqual(settings_str, self.django_default_settings)
        self.assertCountEqual(settings_eval, self.django_default_settings)

        for k, v in self.django_default_settings.items():
            self.assertEqual(v, settings_eval[k])

    @temp_environ()
    def test_update_from_env(self):
        os.environ["DJANGO_SETTINGS_MODULE"] = "settings"
        os.environ["DJANGO_SECRET_KEY"] = 'key'
        os.environ["DJANGO_SITE_ID"] = '255'
        os.environ["DJANGO_EMAIL_PORT"] = '644'
        os.environ["DJANGO_EXTRA_KEY"] = 'asdd'
        os.environ["DJANGO_EXTRA_KEY2"] = 'asdds'
        os.environ["DJANGO_ENGINE_NAME"] = ':memory:'
        os.environ["DJANGO_NOT_ALLOWED"] = "not_allowed"
        os.environ["NOT_ALLOWED"] = "not_allowed"

        import django

        django.setup()

        from django.conf import settings

        self.assertEqual('key', settings.SECRET_KEY)
        self.assertEqual(255, settings.SITE_ID)
        self.assertEqual(25, settings.EMAIL_PORT)
        self.assertEqual('asdd', settings.EXTRA_KEY)
        self.assertIsNone(getattr(settings, "EXTRA_KEY2", None))
        self.assertEqual(getattr(settings, "NOT_ALLOWED", None), None)
        call_command('migrate')


if __name__ == '__main__':
    import unittest
    unittest.main()


tests/settings.py
import os
import sys

from djangobuk_envsettings import update_from_env

update_from_env(
    sys.modules[__name__],
    pre='DJANGO_',
    allowed=[
        'SECRET_KEY',
        'SITE_ID',
    ],
    extra_mapping={
        'EXTRA_KEY': str,
        'EXTRA_KEY2': str,
        'ENGINE_NAME': str,
        'NOT_ALLOWED': str,
    },
    extra_allowed=[
        'EXTRA_KEY',
        'ENGINE_NAME',
    ]
)


DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': ENGINE_NAME,
    }
}


tests/settings_test.py

EXTRA = 2
bad = 1
_SUPER = 3
Super = 5
NEW = "1,2,3,4,5"
a = {}
A = {}


bukdjango_envsettings/__init__.py
import ast

MAPPING = {
    'ABSOLUTE_URL_OVERRIDES': ast.literal_eval,
    'ADMINS': ast.literal_eval,
    'ALLOWED_HOSTS': ast.literal_eval,
    'APPEND_SLASH': ast.literal_eval,
    'AUTHENTICATION_BACKENDS': ast.literal_eval,
    'AUTH_PASSWORD_VALIDATORS': ast.literal_eval,
    'CACHES': ast.literal_eval,
    'CACHE_MIDDLEWARE_SECONDS': ast.literal_eval,
    'CSRF_COOKIE_AGE': ast.literal_eval,
    'CSRF_COOKIE_DOMAIN': ast.literal_eval,
    'CSRF_COOKIE_HTTPONLY': ast.literal_eval,
    'CSRF_COOKIE_SECURE': ast.literal_eval,
    'CSRF_TRUSTED_ORIGINS': ast.literal_eval,
    'CSRF_USE_SESSIONS': ast.literal_eval,
    'DATABASES': ast.literal_eval,
    'DATABASE_ROUTERS': ast.literal_eval,
    'DATA_UPLOAD_MAX_MEMORY_SIZE': ast.literal_eval,
    'DATA_UPLOAD_MAX_NUMBER_FIELDS': ast.literal_eval,
    'DATETIME_INPUT_FORMATS': ast.literal_eval,
    'DATE_INPUT_FORMATS': ast.literal_eval,
    'DEBUG': ast.literal_eval,
    'DEBUG_PROPAGATE_EXCEPTIONS': ast.literal_eval,
    'DISALLOWED_USER_AGENTS': ast.literal_eval,
    'EMAIL_PORT': ast.literal_eval,
    'EMAIL_SSL_CERTFILE': ast.literal_eval,
    'EMAIL_SSL_KEYFILE': ast.literal_eval,
    'EMAIL_TIMEOUT': ast.literal_eval,
    'EMAIL_USE_LOCALTIME': ast.literal_eval,
    'EMAIL_USE_SSL': ast.literal_eval,
    'EMAIL_USE_TLS': ast.literal_eval,
    'FILE_UPLOAD_DIRECTORY_PERMISSIONS': ast.literal_eval,
    'FILE_UPLOAD_HANDLERS': ast.literal_eval,
    'FILE_UPLOAD_MAX_MEMORY_SIZE': ast.literal_eval,
    'FILE_UPLOAD_PERMISSIONS': ast.literal_eval,
    'FILE_UPLOAD_TEMP_DIR': ast.literal_eval,
    'FIRST_DAY_OF_WEEK': ast.literal_eval,
    'FIXTURE_DIRS': ast.literal_eval,
    'FORCE_SCRIPT_NAME': ast.literal_eval,
    'FORMAT_MODULE_PATH': ast.literal_eval,
    'IGNORABLE_404_URLS': ast.literal_eval,
    'INSTALLED_APPS': ast.literal_eval,
    'INTERNAL_IPS': ast.literal_eval,
    'LANGUAGES': ast.literal_eval,
    'LANGUAGES_BIDI': ast.literal_eval,
    'LANGUAGE_COOKIE_AGE': ast.literal_eval,
    'LANGUAGE_COOKIE_DOMAIN': ast.literal_eval,
    'LANGUAGE_COOKIE_HTTPONLY': ast.literal_eval,
    'LANGUAGE_COOKIE_SAMESITE': ast.literal_eval,
    'LANGUAGE_COOKIE_SECURE': ast.literal_eval,
    'LOCALE_PATHS': ast.literal_eval,
    'LOGGING': ast.literal_eval,
    'LOGOUT_REDIRECT_URL': ast.literal_eval,
    'MANAGERS': ast.literal_eval,
    'MIDDLEWARE': ast.literal_eval,
    'MIGRATION_MODULES': ast.literal_eval,
    'NUMBER_GROUPING': ast.literal_eval,
    'PASSWORD_HASHERS': ast.literal_eval,
    'PASSWORD_RESET_TIMEOUT_DAYS': ast.literal_eval,
    'PREPEND_WWW': ast.literal_eval,
    'SECURE_BROWSER_XSS_FILTER': ast.literal_eval,
    'SECURE_CONTENT_TYPE_NOSNIFF': ast.literal_eval,
    'SECURE_HSTS_INCLUDE_SUBDOMAINS': ast.literal_eval,
    'SECURE_HSTS_PRELOAD': ast.literal_eval,
    'SECURE_HSTS_SECONDS': ast.literal_eval,
    'SECURE_PROXY_SSL_HEADER': ast.literal_eval,
    'SECURE_REDIRECT_EXEMPT': ast.literal_eval,
    'SECURE_REFERRER_POLICY': ast.literal_eval,
    'SECURE_SSL_HOST': ast.literal_eval,
    'SECURE_SSL_REDIRECT': ast.literal_eval,
    'SESSION_COOKIE_AGE': ast.literal_eval,
    'SESSION_COOKIE_DOMAIN': ast.literal_eval,
    'SESSION_COOKIE_HTTPONLY': ast.literal_eval,
    'SESSION_COOKIE_SECURE': ast.literal_eval,
    'SESSION_EXPIRE_AT_BROWSER_CLOSE': ast.literal_eval,
    'SESSION_FILE_PATH': ast.literal_eval,
    'SESSION_SAVE_EVERY_REQUEST': ast.literal_eval,
    'SILENCED_SYSTEM_CHECKS': ast.literal_eval,
    'STATICFILES_DIRS': ast.literal_eval,
    'STATICFILES_FINDERS': ast.literal_eval,
    'STATIC_ROOT': ast.literal_eval,
    'STATIC_URL': ast.literal_eval,
    'TEMPLATES': ast.literal_eval,
    'TEST_NON_SERIALIZED_APPS': ast.literal_eval,
    'TIME_INPUT_FORMATS': ast.literal_eval,
    'USE_I18N': ast.literal_eval,
    'USE_L10N': ast.literal_eval,
    'USE_THOUSAND_SEPARATOR': ast.literal_eval,
    'USE_TZ': ast.literal_eval,
    'USE_X_FORWARDED_HOST': ast.literal_eval,
    'USE_X_FORWARDED_PORT': ast.literal_eval,
    'WSGI_APPLICATION': ast.literal_eval,
    'AUTH_USER_MODEL': str,
    'CACHE_MIDDLEWARE_ALIAS': str,
    'CACHE_MIDDLEWARE_KEY_PREFIX': str,
    'CSRF_COOKIE_NAME': str,
    'CSRF_COOKIE_PATH': str,
    'CSRF_COOKIE_SAMESITE': str,
    'CSRF_FAILURE_VIEW': str,
    'CSRF_HEADER_NAME': str,
    'DATETIME_FORMAT': str,
    'DATE_FORMAT': str,
    'DECIMAL_SEPARATOR': str,
    'DEFAULT_CHARSET': str,
    'DEFAULT_EXCEPTION_REPORTER_FILTER': str,
    'DEFAULT_FILE_STORAGE': str,
    'DEFAULT_FROM_EMAIL': str,
    'DEFAULT_INDEX_TABLESPACE': str,
    'DEFAULT_TABLESPACE': str,
    'EMAIL_BACKEND': str,
    'EMAIL_HOST': str,
    'EMAIL_HOST_PASSWORD': str,
    'EMAIL_HOST_USER': str,
    'EMAIL_SUBJECT_PREFIX': str,
    'FILE_CHARSET': str,
    'FORM_RENDERER': str,
    'LANGUAGE_CODE': str,
    'LANGUAGE_COOKIE_NAME': str,
    'LANGUAGE_COOKIE_PATH': str,
    'LOGGING_CONFIG': str,
    'LOGIN_REDIRECT_URL': str,
    'LOGIN_URL': str,
    'MEDIA_ROOT': str,
    'MEDIA_URL': str,
    'MESSAGE_STORAGE': str,
    'MONTH_DAY_FORMAT': str,
    'SECRET_KEY': str,
    'SERVER_EMAIL': str,
    'SESSION_CACHE_ALIAS': str,
    'SESSION_COOKIE_NAME': str,
    'SESSION_COOKIE_PATH': str,
    'SESSION_COOKIE_SAMESITE': str,
    'SESSION_ENGINE': str,
    'SESSION_SERIALIZER': str,
    'SHORT_DATETIME_FORMAT': str,
    'SHORT_DATE_FORMAT': str,
    'SIGNING_BACKEND': str,
    'STATICFILES_STORAGE': str,
    'TEST_RUNNER': str,
    'THOUSAND_SEPARATOR': str,
    'TIME_FORMAT': str,
    'TIME_ZONE': str,  # default None, str
    'X_FRAME_OPTIONS': str,
    'YEAR_MONTH_FORMAT': str,
}


def get_mapping(extra_map: dict = None) -> dict:
    mapping = MAPPING.copy()
    if extra_map:
        mapping.update(extra_map)
    return mapping


def settings_to_dict(settings_module=None) -> dict:
    """
    Convert Django settings module into dict.
    """
    if not settings_module:
        from django.conf import global_settings as settings_module

    return {
        k: type(getattr(settings_module, k)) for k in dir(settings_module) if
        not k.startswith('_') and
        k.upper() == k
    }


def eval_settings(settings: dict, mapping=None) -> dict:
    """
    Given settings map ex. {SECRET_KEY: "asdasd"}
    evaulate values into coresponding Python types.
    """
    mapping = mapping or get_mapping()
    # for opt, value in settings.items():
    #     print(opt, value)
    #     mapping[opt](value)
    return {
        opt: mapping[opt](value) for (opt, value) in settings.items()
    }


manage.py
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys
import django
from djangobuk_envsettings import update_from_env

os.environ["DJANGO_SECRET_KEY"] = 'key'
os.environ["DJANGO_SITE_ID"] = '255'
os.environ["DJANGO_EMAIL_PORT"] = '644'
os.environ["DJANGO_EXTRA_KEY"] = 'asdd'
os.environ["DJANGO_EXTRA_KEY2"] = 'asdds'
os.environ["DJANGO_ENGINE_NAME"] = ':memory:'


def main():
    """Run administrative tasks."""
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'tests.settings')

    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)


if __name__ == '__main__':
    main()
